### PR TITLE
Fix #5643, #5064: rounding issues in floating-point -> integer conversions

### LINKED ISSUE #5643: Right border renders incorrectly when left border is offscreen
- **Sway Version:** 1.5-6991ac8c (Aug 16 2020, branch 'master')

- **Debug Log:** https://gist.github.com/FreeFull/dacb4be28e01048b87b35a19fb878edb

- **Configuration File:** Default sway config.

- **Description:**
When a floating window's left border overlaps with the left edge of the screen or is entirely off the left edge, the right border's rendering will be off by one pixel to the left. I have only tested this with a single monitor. Dragging the top, right or bottom border of the window offscreen doesn't seem to make any of the other borders misrender.

### PR DESCRIPTION
**Pull request by dimitris-triantafyllidis**
Currently, various floating-point expressions involving the coordinates of borders, title bars and content surfaces are directly assigned to integers, and so they are rounded towards zero.

This results in off-by-one distances between these elements when the signs of their coordinates differ.

The glitches are easier to see with semitransparent borders and title bars, where one can see that they can overlap with the content surface when it enters an area with a negative coordinate. 

Fixed by wrapping these expressions with a call to floor before the assignment.

### CODE REVIEW

**Comment by dimitris-triantafyllidis** [comment]
Any feedback on this?

**Comment by emersion** [comment]
>directly assigned to integers, and so they are rounded towards zero.

Doesn't `floor` result in exactly the same thing?

**Comment by dimitris-triantafyllidis** [comment]
For positive floats sure, but negative ones will be rounded towards negative infinity with `floor` instead of towards zero without it. The point of this PR is to round everything towards the same direction to preserve distances when positive and negative coordinates are mixed.

**Comment by dimitris-triantafyllidis** [comment]
@emersion Gentle ping

**Comment by Xyene** [comment]
Hi, sorry for the long review cycle...

Your commit text makes sense and the patch is straightforward, but I'm not sure how to reproduce this locally. Is it necessary for part of the window to be in negative coordinates, and part in positive? Some simple "ELI'm-not-sure-how-to-reproduce-this-in-order-to-test-this-patch" instructions, and maybe some before-after screenshots would be helpful.

**Comment by dimitris-triantafyllidis** [comment]
@Xyene No worries, I understand :-)

Before / after sample:
before
after

In both screenshots the smaller window renders well, but the larger one (which goes partly off the top-left corner of the monitor) has a titlebar, a bottom border, a left border, and a content whose top-left corners have negative coordinates. So the title bar is rendered one pixel off to the right and to the bottom (see overlap with the top edge of the right border), the bottom border is rendered one pixel off to the right, and if you squint a bit you can see that the window content is also rendered one pixel off to the right, overlapping with the right border. This last one is actually not very easy to see in this screenshot...

Note 1: You may have to "shake" the window with the mouse a bit so that the coordinates become fractional, to see the issue.
Note 2: If you try with semi transparent borders and a stacked or tabbed floating container, you may notice that there is a one pixel "jog" between the titlebar(s) and the rest of the window. This is unrelated to the current PR and happens all over the monitor when dragging the window.

If you need further info or screenshots I will be happy to provide them.

**Comment by emersion** [comment]
Thanks a lot of the screenshots, that's very helpful.

Can you rebase on top of master as a single commit? See https://git-rebase.io/.

**Comment by dimitris-triantafyllidis** [comment]
I resolved a conflict here and the commits became two, then I just redid the branch locally and force-pushed as a single commit, am I missing something?

**Review by emersion — APPROVED** [review]
Oh, sorry, I didn't notice you pushed a new commit.

This LGTM, thanks. I'll let @Xyene merge this if he agrees.

**Review by Xyene — APPROVED** [review]
Thanks for working on this!

**Pull request merged**
